pyccel/stdlib/cwrapper/cwrapper_ndarrays.c

```diff
@@ -72,3 +199,77 @@ t_ndarray	pyarray_to_c_ndarray(PyArrayObject *a)
 
 	return array;
 }
+
+
+/*
+ * Function: pyarray_check
+ * --------------------
+ * Check Python Object (DataType, Rank, Order):
+ *
+ * 	Parameters	:
+ *		a 	  : python array object
+ *      dtype : desired data type enum
+ *		rank  : desired rank
+ *		flag  : desired order flag
+ * 	Returns		:
+ *		return true if no error occurred otherwise it will return false
+ */
+bool	pyarray_check(PyArrayObject *o, int dtype, int rank, int flag)
+{
+	if (!_check_pyarray_type((PyObject *)o)) return false;
+
+	// check array element type / rank / order
+	if(!check_pyarray_dtype(o, dtype)) return false;
+
+	if(!_check_pyarray_rank(o, rank)) return false;
+
+	if(rank > 1 && !_check_pyarray_order(o, flag)) return false;
+
+	return true;
+}
+
+
+/*
+ * Function: nd_ndim
+ * --------------------
+ * Return the shape in the n dimension.
+ *
+ * 	Parameters	:
+ *		a 	  : python array object
+ *      index : dimension index
+ * 	Returns		:
+ *		return 0 if object is NULL or shape at indexed dimension
+ * reference of the used c/numpy api function
+ * -------------------------------------------
+ * https://numpy.org/doc/1.17/reference/c-api.array.html#c.PyArray_DIM
+ */
+int     nd_ndim(t_ndarray *a, int n)
+{
+	if (a == NULL)
+		return 0;
+
```

Comment:
I think we have to raise `index out of range` error in `n >= array.nd`

Reply:
I don't think it is possible to have `n >= array.nd`. This function is only called by pyccel to pass an ndarray to a fortran function by doing (for example):
```
bind_c_array_int32_1d_scalar_add(nd_ndim(&x, 0), nd_data(&x), a);
```
When the array `x` is collected its rank is checked. We will only ask for indices where `n<x.rank` and as this is checked there is no danger of out of range problems

Reply:
hmm, I tried to reproduce the error for that case but I get  a different error:
```python
import numpy as np
a = np.array([[1, 2], [3, 4]])
n = 2
print(a.shape[n])
```
I get 
```bash
TypeError: tuple indices must be integers or slices, not variable
```

Reply:
This will not reproduce the error above as the function above is only used in the wrapper (hence it appearing in `cwrapper_ndarrays.c`). I don't think it is possible to reproduce the error which is why I don't think it needs handling. It would appear if you passed a 1d array to a function translated to fortran which expects a 2d array, if you comment the section that tests if the array that was passed is indeed a 2d array

Reply:
yap, until now we are dealing only with user mistakes, what do you think in case of someone of us modified some part of the wrapper and get a `segfault` error, how much time does he need to figure out that problem?

Reply:
Hmm, I don't know. Given that the function is only used in 1 line, if modifications are made to the code incrementally it should be reasonably easy to see where in the generated code you need to look and it should be fairly obvious that you're doing something stupid. This line is unlikely to be touched often. As the size has already been tested I am not convinced that the trade-off between catching very stupid developer behaviour and having 2 tests of the size is worth it. I'm also struggling to see how you can make this mistake.
The function is used [here](https://github.com/pyccel/pyccel/blob/wrapper_part_2/pyccel/codegen/printing/cwrappercode.py#L279)
so you would have to replace `rank` in:
```
[FunctionCall(array_get_dim, [arg_address, i]) for i in range(argument.rank)]
```
with a `LiteralInteger` or a sum or something very strange like that

Reply:
yes, you are right.